pkg/rpc/server.go

```diff
@@ -115,39 +124,145 @@ func unexport(volumeName string) error {
 	return nil
 }
 
-func unmount(volumeName string) error {
-	logrus.Infof("Unmounting volume %v", volumeName)
-
-	mountPath := types.GetMountPath(volumeName)
+func (s *ShareManagerServer) unmount(vol volume.Volume) error {
+	mountPath := types.GetMountPath(vol.Name)
 
 	mounter := mount.New("")
 	notMounted, err := mount.IsNotMountPoint(mounter, mountPath)
+	if err != nil {
+		return errors.Wrapf(err, "failed to check mount point %v", mountPath)
+	}
 	if notMounted {
-		return err
+		return nil
 	}
 
 	return volume.UnmountVolume(mountPath)
 }
 
 func (s *ShareManagerServer) Unmount(ctx context.Context, req *empty.Empty) (resp *empty.Empty, err error) {
-	volumeName := s.manager.GetVolumeName()
+	vol := s.manager.GetVolume()
+	if vol.Name == "" {
+		s.logger.Warn("Volume name is missing")
+		return &empty.Empty{}, nil
+	}
+
+	log := s.logger.WithField("volume", vol.Name)
+
+	if !nfsServerIsRunning() {
+		log.Info("NFS server is not running, skip unexporting and unmounting volume")
+		return &empty.Empty{}, nil
+	}
+
+	// Blindly mark the volume as unexported, even if the unmount fails.
+	// Mount() will re-export the volume and mark it as exported if needed.
+	s.manager.SetShareExported(false)
 
 	defer func() {
 		if err != nil {
-			logrus.WithError(err).Errorf("Failed to unmount volume %v", volumeName)
+			log.WithError(err).Errorf("Failed to unexport and unmount volume")
 		}
 	}()
 
-	err = unexport(volumeName)
+	log.Info("Unexporting volume")
+	err = s.unexport(vol)
 	if err != nil {
 		return nil, grpcstatus.Error(grpccodes.Internal, err.Error())
 	}
 
-	err = unmount(volumeName)
+	log.Info("Unmounting volume")
+	for i := 0; i < unmountRetryCount; i++ {
+		err := s.unmount(vol)
+		if err != nil {
```

Comment:
If err is nil, we should break from the forloop.

Reply:
Updated. Thank you.